bug:TriggersAnnotationEnricher should not execute in PlatformMode.kubernetes

AmitPatil05 opened the pull request:
## Description

Fixes #2559

Added a check in the enriching process to only do it if the Kubernetes environment is OpenShift. Also added 2 new unit tests and updated releevant unit tests.

## Type of change
<!---
What types of changes does your code introduce? Put an `x` in all the boxes that apply
-->
 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [ ] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] I have read the [contributing guidelines](https://www.eclipse.dev/jkube/contributing)
 - [x] I signed-off my commit with a user that has signed the [Eclipse Contributor Agreement](https://www.eclipse.org/legal/ECA.php)
 - [ ] My code follows the style guidelines of this project
 - [ ] I Keep It Small and Simple: The smaller the PR is, the easier it is to review and have it merged
 - [ ] I use [conventional commits](https://www.conventionalcommits.org/) in my commit messages
 - [ ] I have performed a self-review of my code
 - [ ] I Added [CHANGELOG](../CHANGELOG.md) entry
 - [ ] I have updated the [documentation](../kubernetes-maven-plugin/doc) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=jkubeio_jkube) report
 - [ ] I have added tests that prove my fix is effective or that my feature works
 - [ ] New and existing unit tests pass locally with my changes
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

<!--
Integration tests (https://github.com/jkubeio/jkube-integration-tests)
Please check integration tests and provide/improve tests if necessary.

Open your PR in Draft mode and verify all of the applicable Checklist items before marking your issue as ready
-->

rohanKanojia commented on jkube-kit/enricher/generic/src/test/java/org/eclipse/jkube/enricher/generic/TriggersAnnotationEnricherTest.java:
```diff
@@ -55,6 +56,34 @@ void setUp() {
 
+    @Test
+    void testEnrichmentNotPerformedInNonOpenShiftMode() throws IOException {
+        // Create a mock KubernetesListBuilder
```
If you want to add a description maybe use `@DisplayName` annotation

rohanKanojia commented on jkube-kit/enricher/generic/src/test/java/org/eclipse/jkube/enricher/generic/TriggersAnnotationEnricherTest.java:
```diff
@@ -55,6 +56,34 @@ void setUp() {
+    @Test
+    void testEnrichmentPerformedInOpenShiftMode() throws IOException {
+        // Create a mock KubernetesListBuilder
+        KubernetesListBuilder builder = mock(KubernetesListBuilder.class);
```
Could you please adapt these tests like existing tests in this class? You're using way too much mocking.

Try to create a real KubernetesListBuilder with some valid Kubernetes object. For this case verify that the object contains the annotation after `enricher.enrich` gets called. For `testEnrichmentNotPerformedInNonOpenShiftMode` you can verify that object does not contain that annotation

AmitPatil05 commented:
Hello, 
As per your comment, I have created a real KubernetesListBuilder with a valid Kubernetes object (StatefulSet).

1. In testEnrichmentNotPerformedInNonOpenShiftMode case, I called enrich method with PlatformMode.kubernetes. Verification was done in such a way that, the StatefulSet is retrieved from the builder, and assertions are made to verify the presence of the original annotation ("annkey": "annvalue") and the absence of any additional annotations, ensuring that only the original annotation remains unchanged by using hasSize(1).
2. In testEnrichmentPerformedInOpenShiftMode case, I called enrich method with PlatformMode.openshift. Verification was done in such a way that, the original annotation stays ("annkey": "annvalue") and the addition of a new annotation ("image.openshift.io/triggers") related to OpenShift triggers is done. Ensured that the total number of annotations being 2 using hasSize(2), which indicates that both the original and new annotations are present.


Can you please review and tell if this approach is fine? 


Thanks,
Amit

AmitPatil05 commented:
Sonar build is failing because of [INFO] Kubernetes Maven Plugin :: Integration Tests ....... FAILURE [04:17 min]

I am not able to find the root cause of this. The change is just in test. Maybe is it because of the change in PlatformMode to openshift in the test class? Also it might be because of now enricher is specifically using openshift as platformMode and not anything else. 

Can you please guide me here?

rohanKanojia commented:
@AmitPatil05 : Your changes now don't execute in kubernetes mode. Some integration testshave wrong expectations. You need to remove this annotation from `expected/kubernetes.yml` 

https://github.com/eclipse/jkube/blob/62080bfd0ce8c65245fef69723c95816d49661bb/kubernetes-maven-plugin/it/src/it/dependency-resources/expected/kubernetes.yml#L75

rohanKanojia commented:
@AmitPatil05 : Kubernetes Integration tests seem to be fixed now.

However, I see some occurrences of this annotation in `gradle-plugin/it` module as well:
```
src/it/controller/expected/daemonset/kubernetes.yml:12:        image.openshift.io/triggers: "@ignore@"
src/it/controller/expected/replicaset/kubernetes.yml:12:        image.openshift.io/triggers: "@ignore@"
src/it/controller/expected/statefulset/kubernetes.yml:12:        image.openshift.io/triggers: "@ignore@"
src/it/triggersannotation/expected/kubernetes.yml:12:      image.openshift.io/triggers: |-
```

AmitPatil05 commented:
Hi @rohanKanojia 
I removed all the annotation from kubernetes.yaml files in 7da3367 file. The Sonar build failed. Then I put back the annotations in some of the files in the latest commit. Still the Sonar build failed. Now the Kubernetes gradle plugin build is successful but integrationn test is now failing. Does this means some of the unit tests are failing right now? Is there something wrong with what I implemented in the test?

rohanKanojia commented on gradle-plugin/it/src/it/triggersannotation/expected/kubernetes.yml:
```diff
@@ -9,7 +9,6 @@ items:
       jkube.eclipse.org/git-url: "@ignore@"
       jkube.eclipse.org/git-commit: "@ignore@"
       jkube.eclipse.org/git-branch: "@ignore@"
-      image.openshift.io/triggers: |-
```
@AmitPatil05 You haven't removed this correctly. This attribute has multi line scalar value

You need to delete till line 19 (where json array ends)

AmitPatil05 commented:
Hi @rohanKanojia , I removed the annotations correctly. Can you please review the latest changes?

rohanKanojia commented on jkube-kit/enricher/generic/src/test/java/org/eclipse/jkube/enricher/generic/TriggersAnnotationEnricherTest.java:
```diff
@@ -55,6 +55,56 @@ void setUp() {
+
+        StatefulSet res = (StatefulSet) builder.build().getItems().get(0);
+        assertThat(res.getMetadata().getAnnotations())
+            .containsEntry("annkey", "annvalue")
```
```suggestion
            .containsEntry("annkey", "annvalue")
            .doesNotContainKey("image.openshift.io/triggers")
```

rohanKanojia commented:
@AmitPatil05 : Please squash your commits into one commit.

manusa reviewed (approved):
LGTM, thx!
